redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+            rediss://[[username]:[password]]@localhost:6379/0
+            unix://[[username]:[password]]@/path/to/socket.sock?db=0
+
+        Three URL schemes are supported:
+
+        - `redis://` creates a TCP socket connection. See more at:
+          <https://www.iana.org/assignments/uri-schemes/prov/redis>
+        - `rediss://` creates a SSL wrapped TCP socket connection. See more at:
+          <https://www.iana.org/assignments/uri-schemes/prov/rediss>
+        - ``unix://``: creates a Unix Domain Socket connection.
+
+        The username, password, hostname, path and all querystring values
+        are passed through urllib.parse.unquote in order to replace any
+        percent-encoded values with their corresponding characters.
+
+        There are several ways to specify a database number. The first value
+        found will be used:
+            1. A ``db`` querystring option, e.g. redis://localhost?db=0
+            2. If using the redis:// or rediss:// schemes, the path argument
+               of the url, e.g. redis://localhost/0
+            3. A ``db`` keyword argument to this function.
+
+        If none of these options are specified, the default db=0 is used.
+
+        All querystring options are cast to their appropriate Python types.
+        Boolean arguments can be specified with string values "True"/"False"
+        or "Yes"/"No". Values that cannot be properly cast cause a
+        ``ValueError`` to be raised. Once parsed, the querystring arguments
+        and keyword arguments are passed to the ``ConnectionPool``'s
+        class initializer. In the case of conflicting arguments, querystring
+        arguments always win.
+
+        """
+        return cls(url=url, **kwargs)
+
+    def on_connect(self, connection):
+        """
+        Initialize the connection, authenticate and select a database and send
+         READONLY if it is set during object initialization.
+        """
+        connection.set_parser_class(ClusterParser)
+        connection.on_connect()
+
+        if self.read_from_replicas:
+            # Sending READONLY command to server to configure connection as
+            # readonly. Since each cluster node may change its server type due
+            # to a failover, we should establish a READONLY connection
+            # regardless of the server type. If this is a primary connection,
+            # READONLY would not affect executing write commands.
+            connection.send_command('READONLY')
+            if str_if_bytes(connection.read_response()) != 'OK':
+                raise ConnectionError('READONLY command failed')
+
+    def get_redis_connection(self, node):
+        if not node.redis_connection:
+            self.nodes_manager.create_redis_connections([node])
+        return node.redis_connection
+
+    def get_node(self, node_name=None, host=None, port=None):
+        return self.nodes_manager.get_node(node_name, host, port)
+
+    def get_all_primaries(self):
```

Comment:
s/get_all_primaryies/get_primaries

I don't think we're going to want only some of the primaries. If someone does.. they can me a slice.

Reply:
changed